dev/autobuildsheriff/sherrif.go

```diff
@@ -0,0 +1,147 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/buildkite/go-buildkite/v3/buildkite"
+)
+
+type sherrifOptions struct {
+	FailuresThreshold int
+	BuildTimeout      time.Duration
+}
+
+type commitInfo struct {
+	Commit string
+	Author string
+}
+
+type sherrifResults struct {
+	Locked        bool
+	LockModified  bool
+	FailedCommits []commitInfo
+}
+
+// buildsherrif is the main sherrifing program. It checks the given builds for relevant
+// failures and runs lock/unlock operations on the given branch.
+func buildsherrif(ctx context.Context, branch branchLocker, builds []buildkite.Build, opts sherrifOptions) (results *sherrifResults, err error) {
+	results = &sherrifResults{}
+
+	// Scan for first build with a meaningful state
+	var firstFailedBuild int
+	for i, b := range builds {
+		if isBuildPassed(b) {
+			fmt.Printf("most recent finished build %d passed\n", *b.Number)
+			results.LockModified, err = branch.Unlock(ctx)
+			if err != nil {
+				return nil, fmt.Errorf("unlockBranch: %w", err)
+			}
+			return
+		}
+		if isBuildFailed(b, opts.BuildTimeout) {
+			fmt.Printf("most recent finished build %d failed\n", *b.Number)
+			firstFailedBuild = i
+			break
+		}
+
+		// Otherwise, keep looking for builds
```

Comment:
Very minor suggestion to ease reading that code later on (I think that whether or not cancelled builds are taken in account is a question that's going to pop in the the reader's mind). 

```suggestion
		// Otherwise, keep looking for builds, ignoring cancelled or running builds. 
```

Reply:
Builds should never be cancelled on main - I will add those to the failure condition

Reply:
Changed docstring to:

```
Otherwise, keep looking for a completed (failed or passed) build
```